### DISCUSSION TITLE
Handle different runs concurrently on multiple tracking servers?

### DISCUSSION THREAD

**Discussion by browser-bug** [mlflow/mlflow#5830] [Q&A] [answered]
Hi everyone! Are there any resources or insights on handling multiple tracking servers concurrently? We're trying to deploy a simple FastAPI that basically launches, potentially concurrently, multiple runs on different Tracking Servers using the MLflow Python API. We've seen that there's no clear way to explicitly assign the Tracking URI during the `mlflow.projects.run` function and so we're obliged to use `set_tracking_uri` every time before launching the new run (which I quote "does not affect the currently active run (if one exists), but takes effect for successive runs."). Problem is that it may happens that multiple runs go in conflict between each other and some random errors like `mlflow.exceptions.RestException: RESOURCE_DOES_NOT_EXIST` may occur.
Is there a way to handle this use case scenario or MLflow is still too unripe to be handling multiple tracking servers on a single endpoint like this?

**Comment by dbczumar** [comment, marked as answer]
Hi @browser-bug , use of `set_tracking_uri()` prior to `mlflow.projects.run()` is indeed the recommended approach for launching MLflow Project runs against multiple workspaces. Note that, once you have obtained the resulting run ID, you can access the run via  `mlflow.tracking.MlflowClient.get_run()` (https://mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.get_run); the constructor of `MlflowClient` accepts a Tracking URI (https://mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient). This approach may make it easier to avoid `RESOURCE_DOES_NOT_EXIST` errors.
